### PR TITLE
Utilities: Refactor and improve du

### LINKED ISSUE #12899: Utilities: du's summarize flag incorrectly reports zero bytes all the time
In this case, $PWD has dozens of files, but with `-s` it gets reported as zero. Without -s, it gets correctly reported as containing over 36 MiB of files. (the size is supposed to change with every command run, so pay no attention there :) ).

du-bug

I'm not convinced this ever worked, to be honest.

### PR DESCRIPTION
**Pull request by SeekingBlues**
* Make sure the size calculation is correct even if `--max-depth` or `--threshold` is specified.
* Correctly implement apparent size calculation (`st_size`).
* Improve POSIX correctness by using a default block size of 512 bytes and including the size of directories themselves.
* Add more options.
* Clarify some documentation.

Fixes #12899.

### CODE REVIEW

**Review by timschumi — CHANGES_REQUESTED** [review]
Coincidentally, this also checks off a point in #12303.

That said, this is way too much for a single commit. Please split it up into separate commits, which should also make it less impossible to review.

**Comment by SeekingBlues** [comment]
Sorry, I totally forgot about splitting the commits. Hopefully this looks better. Let me know if there is anything that I can improve.

File: Userland/Utilities/du.cpp [hunk 1]
```diff
@@ -110,22 +110,26 @@ ErrorOr<void> parse_args(Main::Arguments const& arguments)
...
 {
-    struct stat path_stat = TRY(Core::System::lstat(path.characters()));
-    off_t directory_size = 0;
-    bool const is_directory = S_ISDIR(path_stat.st_mode);
-    if (--max_depth >= 0 && is_directory) {
-        auto di = Core::DirIterator(path, Core::DirIterator::SkipParentAndBaseDir);
+    auto path_stat = TRY(Core::System::lstat(path.characters()));
+    bool is_directory = S_ISDIR(path_stat.st_mode);
+    off_t size = 0;
+    if (is_directory) {
```
Comment (Hendiadyoin1) [inline Userland/Utilities/du.cpp:138]: The check against the maximum recursion depth seems to be missing now

File: Userland/Utilities/du.cpp [hunk 2]
```diff
@@ -144,8 +139,7 @@ ErrorOr<off_t> print_space_usage(String const& path, unsigned depth)
         out("{}", human_readable_size(size));
     } else {
         constexpr long long block_size = 1024;
-        size = size / block_size + (size % block_size != 0);
-        out("{}", size);
+        out("{}", howmany(size, block_size));
```
Comment (Hendiadyoin1) [inline Userland/Utilities/du.cpp]: `ceil_div` from `AK/StdLibExtras.h` might be even better.

File: Userland/Utilities/du.cpp [hunk 3]
```diff
@@ -18,7 +18,7 @@ struct DuOption {
         NotUsed,
         Modification,
         Access,
-        Status
+        Change
```
Comment (Hendiadyoin1) [inline Userland/Utilities/du.cpp:21]: This seems like an unrelated change.

File: Userland/Utilities/du.cpp [hunk 4]
```diff
@@ -132,36 +132,35 @@ ErrorOr<off_t> print_space_usage(String const& path, unsigned depth)
...
-        constexpr long long block_size = 512;
-        out("{}", howmany(size, block_size));
-    }
+    bool in_threshold = s_option.threshold == 0 || ((s_option.threshold > 0 && size < s_option.threshold) || (s_option.threshold < 0 && size > -s_option.threshold));
+    if ((is_directory || (s_option.all || depth == 0)) && depth <= s_option.max_depth && in_threshold) {
+        if (s_option.human_readable) {
+            out("{}", human_readable_size(size));
+        } else {
+            constexpr long long block_size = 512;
+            out("{}", howmany(size, block_size));
```
Comment (Hendiadyoin1) [inline Userland/Utilities/du.cpp]: -> `ceil_div`

**Pull request closed without merging**